tests/test/functions.rs

```diff
@@ -1,5 +1,121 @@
 use super::*;
 
+#[test]
+fn function_implement_fn_traits() {
+    test! {
+        program {
+            #[lang(fn_once)]
+            trait FnOnce<Args> {
+                type Output;
+            }
+
+            #[lang(fn_mut)]
+            trait FnMut<Args> where Self: FnOnce<Args> { }
+
+            #[lang(fn)]
+            trait Fn<Args> where Self: FnMut<Args> { }
+        }
+
+        goal {
+            fn(u8): FnOnce<(u8,)>
+        } yields {
+            "Unique; substitution [], lifetime constraints []"
+        }
+
+        goal {
+            fn(u8): FnMut<(u8,)>
+        } yields {
+            "Unique; substitution [], lifetime constraints []"
+        }
+
+        goal {
+            fn(u8): Fn<(u8,)>
+        } yields {
+            "Unique; substitution [], lifetime constraints []"
+        }
+
+        goal {
+            Normalize(<fn(u8) as FnOnce<(u8,)>>::Output -> ())
```

Comment:
You *could* just combine this into the goal above like `fn(u8): FnOnce<(u8,), Output = ()>`, but since there's a couple output-specific test goals, this is probably fine.

Reply:
Side note: it would be nice to have a short comment for each of these goals to actually state *why* this specific test. It'll be important when we add `FnDef` and `Closure` to decide what goals *also* need to be added there, and which *don't*